lib/socket.c

```diff
@@ -568,98 +568,101 @@ iscsi_read_from_socket(struct iscsi_context *iscsi)
-		return -1;
-	}
-	if (data_size != 0) {
-		unsigned char padding_buf[3];
-		unsigned char *buf = padding_buf;
-		struct scsi_iovector * iovector_in;
-
-		count = data_size - in->data_pos;
-
-		/* first try to see if we already have a user buffer */
-		iovector_in = iscsi_get_scsi_task_iovector_in(iscsi, in);
-		if (iovector_in != NULL && count > padding_size) {
-			uint32_t offset = scsi_get_uint32(&in->hdr[40]);
-			count = iscsi_iovector_readv_writev(iscsi, iovector_in, in->data_pos + offset, count - padding_size, 0);
-		} else {
-			if (iovector_in == NULL) {
-				if (in->data == NULL) {
-					in->data = iscsi_malloc(iscsi, data_size);
+		if (data_size < 0 || data_size > (ssize_t)iscsi->initiator_max_recv_data_segment_length) {
+			iscsi_set_error(iscsi, "Invalid data size received from target (%d)", (int)data_size);
+			return -1;
+		}
+		if (data_size != 0) {
+			unsigned char padding_buf[3];
+			unsigned char *buf = padding_buf;
+			struct scsi_iovector * iovector_in;
+
+			count = data_size - in->data_pos;
+
+			/* first try to see if we already have a user buffer */
+			iovector_in = iscsi_get_scsi_task_iovector_in(iscsi, in);
+			if (iovector_in != NULL && count > padding_size) {
+				uint32_t offset = scsi_get_uint32(&in->hdr[40]);
+				count = iscsi_iovector_readv_writev(iscsi, iovector_in, in->data_pos + offset, count - padding_size, 0);
+			} else {
+				if (iovector_in == NULL) {
 					if (in->data == NULL) {
-						iscsi_set_error(iscsi, "Out-of-memory: failed to malloc iscsi_in_pdu->data(%d)", (int)data_size);
-						return -1;
+						in->data = iscsi_malloc(iscsi, data_size);
+						if (in->data == NULL) {
+							iscsi_set_error(iscsi, "Out-of-memory: failed to malloc iscsi_in_pdu->data(%d)", (int)data_size);
+							return -1;
+						}
 					}
+					buf = &in->data[in->data_pos];
 				}
-				buf = &in->data[in->data_pos];
+				count = recv(iscsi->fd, buf, count, 0);
 			}
-			count = recv(iscsi->fd, buf, count, 0);
-		}
-		
-		if (count == 0) {
-			return -1;
-		}
-		if (count < 0) {
-			if (errno == EINTR || errno == EAGAIN) {
-				return 0;
+			
+			if (count == 0) {
+				/* remote side has closed the socket. */
+				return -1;
+			}
+			if (count < 0) {
+				if (errno == EINTR || errno == EAGAIN) {
+					break;
+				}
+				iscsi_set_error(iscsi, "read from socket failed, "
+						"errno:%d %s", errno,
+						iscsi_get_error(iscsi));
+				return -1;
 			}
-			iscsi_set_error(iscsi, "read from socket failed, "
-					"errno:%d %s", errno,
-					iscsi_get_error(iscsi));
-			return -1;
-		}
-
-		in->data_pos += count;
-	}
 
-	if (in->data_pos < data_size) {
-		return 0;
-	}
+			in->data_pos += count;
+		}
 
-	ISCSI_LIST_ADD_END(&iscsi->inqueue, in);
-	iscsi->incoming = NULL;
+		if (in->data_pos < data_size) {
+			break;
+		}
 
+		ISCSI_LIST_ADD_END(&iscsi->inqueue, in);
+		iscsi->incoming = NULL;
+	} while (iscsi->is_loggedin && iscsi->tcp_nonblocking);
```

Comment:
At the end of this loop there is opportunity for a tiny optimization.

If the length of iscsi->waitpdu and iscsi->incoming are the same then except for any target initiated NOPs we should have a received any and all possible pdus from this socket and can abort early.

This avoids running the loop one more time just to fail with EAGAIN at the recs/readv. Just avoiding that recv/readv syscall will shave at least 10us off this function and thus the latency.